### PR TITLE
Moving definitions to a dictionary

### PR DESCRIPTION
**Pull request by jeff-allen-mongo**
As discussed, I've updated some of the typing in the functions as well as moved the logic for the definitions from being a list of lists to a dictionary.

### CODE REVIEW

File: mut/redirects/redirect_main.py [hunk 1]
```diff
@@ -21,18 +21,19 @@ class RedirectContext:
     def __init__(self) -> None:
         self.rules = []  # type: List[RuleDefinition]
         self.symlinks = []  # type: List[List[str]]
-        self.definitions = []  # type: List[List[str]]
+        self.definitions = {}  # type: dict
+        return None
```
Comment (i80and) [inline mut/redirects/redirect_main.py]: Never return a value from a constructor

File: mut/redirects/redirect_main.py [hunk 2]
```diff
@@ -21,18 +21,19 @@ class RedirectContext:
     def __init__(self) -> None:
         self.rules = []  # type: List[RuleDefinition]
         self.symlinks = []  # type: List[List[str]]
-        self.definitions = []  # type: List[List[str]]
+        self.definitions = {}  # type: dict
```
Comment (i80and) [inline mut/redirects/redirect_main.py]: This should be e.g. `Dict[str, str]`

File: mut/redirects/redirect_main.py [hunk 3]
```diff
@@ -21,18 +21,19 @@ class RedirectContext:
...
+        return None
 
     def add_definition(self, key: str, value: str) -> None:
-        d = [key, value]
-        self.definitions.append(d)
+        self.definitions[key] = value
+        return None
```
Comment (i80and) [inline mut/redirects/redirect_main.py]: `None` is the implicit return type; no need to return None explicitly

File: mut/redirects/redirect_main.py [hunk 4]
```diff
@@ -57,6 +58,7 @@ def generate_rule(self, is_temp: bool, version: str, old_url: str, new_url: str,
                     self.generate_rule(is_temp, symlink[0], old_url, new_url, True)
 
         self.rules.append(new_rule)
+        return None
```
Comment (i80and) [inline mut/redirects/redirect_main.py]: See above

File: mut/redirects/redirect_main.py [hunk 5]
```diff
@@ -104,11 +101,13 @@ def write_to_file(rules: List[RuleDefinition], output_path: str) -> None:
             f.write(line)
             f.write('\n')
         f.close()
+    return None
```
Comment (i80and) [inline mut/redirects/redirect_main.py]: See above

File: mut/redirects/redirect_main.py [hunk 6]
```diff
@@ -122,36 +121,36 @@ def parse_source_file(source_path: str, output: str) -> None:
...
+                            value = value + type_split[x] + ' '
                     else:
                         value = type_split[2]
 
+                    value = value.strip()
                     rc.add_definition(key, value)
                     # see if any definitions have to be substituted with other definitions:
-                    for x in range(0, len(rc.definitions)):
-                        sub_regex = '{(.*?)}'
+                    for key, value in rc.definitions.items():
```
Comment (i80and) [inline mut/redirects/redirect_main.py]: Refer to our verbal discussion wrt substitution :D

File: mut/redirects/redirect_main.py (hunk 1 above)
Comment (jeff-allen-mongo) [reply mut/redirects/redirect_main.py]: Oops :no_mouth:

**Pull request merged**